tools/pytorch2onnx.py

```diff
@@ -1,125 +1,144 @@
 import argparse
-import io
+import os.path as osp
+from functools import partial
 
 import mmcv
+import numpy as np
 import onnx
+import onnxruntime as rt
 import torch
 from mmcv.runner import load_checkpoint
-from onnx import optimizer
-from torch.onnx import OperatorExportTypes
+from onnx_util.symbolic import register_extra_symbolics
 
 from mmdet.models import build_detector
-from mmdet.ops import RoIAlign, RoIPool
-
-
-def export_onnx_model(model, inputs, passes):
-    """
-    Trace and export a model to onnx format.
-    Modified from https://github.com/facebookresearch/detectron2/
-
-    Args:
-        model (nn.Module):
-        inputs (tuple[args]): the model will be called by `model(*inputs)`
-        passes (None or list[str]): the optimization passed for ONNX model
-
-    Returns:
-        an onnx model
-    """
-    assert isinstance(model, torch.nn.Module)
-
-    # make sure all modules are in eval mode, onnx may change the training
-    # state of the module if the states are not consistent
-    def _check_eval(module):
-        assert not module.training
-
-    model.apply(_check_eval)
-
-    # Export the model to ONNX
-    with torch.no_grad():
-        with io.BytesIO() as f:
-            torch.onnx.export(
-                model,
-                inputs,
-                f,
-                operator_export_type=OperatorExportTypes.ONNX_ATEN_FALLBACK,
-                # verbose=True,  # NOTE: uncomment this for debugging
-                # export_params=True,
-            )
-            onnx_model = onnx.load_from_string(f.getvalue())
-
-    # Apply ONNX's Optimization
-    if passes is not None:
-        all_passes = optimizer.get_available_passes()
-        assert all(p in all_passes for p in passes), \
-            f'Only {all_passes} are supported'
-    onnx_model = optimizer.optimize(onnx_model, passes)
-    return onnx_model
+
+
+def pytorch2onnx(model,
+                 input_img,
+                 input_shape,
+                 opset_version=11,
+                 show=False,
+                 output_file='tmp.onnx',
+                 verify=False):
+    model.cpu().eval()
+    # use dummy input to execute model for tracing
+    one_img = mmcv.imread(input_img, 'color')
+    one_img = mmcv.imresize(one_img, input_shape[2:]).transpose(2, 0, 1)
+    # normalize the input images
+    one_img = torch.from_numpy((one_img - 128) / 256).unsqueeze(0).float()
+    (_, C, H, W) = input_shape
+    one_meta = {
+        'img_shape': (H, W, C),
```

Comment:
When input shape is not the same as input image, will the result be incorrect?

Reply:
> When input shape is not the same as input image, will the result be incorrect?

It will be incorrect. As the meta data will be used when calling bbox_head.get_bboxes, and the img_shape in meta data is used when calling delta2bbox. All data in meta data, which have been used while executing, will be regarded as hyper-parameter in ONNX. We have to make sure they are correct.

Reply:
We may also add these in the comments and docs.